internal/app/orchestrator/orchestrator_test.go

```diff
@@ -2,19 +2,99 @@ package orchestrator
+	"time"
 
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	gcp "github.com/envoyproxy/go-control-plane/pkg/cache/v2"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/golang/protobuf/ptypes/duration"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/envoyproxy/xds-relay/internal/app/cache"
 	"github.com/envoyproxy/xds-relay/internal/app/mapper"
 	"github.com/envoyproxy/xds-relay/internal/app/upstream"
 	upstream_mock "github.com/envoyproxy/xds-relay/internal/app/upstream/mock"
 	"github.com/envoyproxy/xds-relay/internal/pkg/log"
+	"github.com/envoyproxy/xds-relay/internal/pkg/util/yamlproto"
+	aggregationv1 "github.com/envoyproxy/xds-relay/pkg/api/aggregation/v1"
 	bootstrapv1 "github.com/envoyproxy/xds-relay/pkg/api/bootstrap/v1"
-	"github.com/golang/protobuf/ptypes/duration"
+)
 
-	aggregationv1 "github.com/envoyproxy/xds-relay/pkg/api/aggregation/v1"
+type mockSimpleUpstreamClient struct {
+	responseChan <-chan *v2.DiscoveryResponse
+}
 
-	"github.com/stretchr/testify/assert"
-)
+func (m mockSimpleUpstreamClient) OpenStream(req v2.DiscoveryRequest) (<-chan *v2.DiscoveryResponse, func(), error) {
+	return m.responseChan, func() {}, nil
+}
+
+type mockMultiStreamUpstreamClient struct {
+	ldsResponseChan <-chan *v2.DiscoveryResponse
+	cdsResponseChan <-chan *v2.DiscoveryResponse
+
+	t      *testing.T
+	mapper mapper.Mapper
+}
+
+func (m mockMultiStreamUpstreamClient) OpenStream(
+	req v2.DiscoveryRequest,
+) (<-chan *v2.DiscoveryResponse, func(), error) {
+	aggregatedKey, err := m.mapper.GetKey(req)
+	assert.NoError(m.t, err)
+
+	if aggregatedKey == "lds" {
+		return m.ldsResponseChan, func() {}, nil
+	} else if aggregatedKey == "cds" {
+		return m.cdsResponseChan, func() {}, nil
+	}
+
+	m.t.Errorf("Unsupported aggregated key, %s", aggregatedKey)
+	return nil, func() {}, nil
+}
+
+func newMockOrchestrator(t *testing.T, mapper mapper.Mapper, upstreamClient upstream.Client) *orchestrator {
+	orchestrator := &orchestrator{
+		logger:         log.New("info"),
+		mapper:         mapper,
+		upstreamClient: upstreamClient,
+		downstreamResponseMap: downstreamResponseMap{
+			responseChannels: make(map[*gcp.Request]chan gcp.Response),
+		},
+		upstreamResponseMap: upstreamResponseMap{
+			responseChannels: make(map[string]upstreamResponseChannel),
+		},
+	}
+
+	cache, err := cache.NewCache(1000, orchestrator.onCacheEvicted, 10*time.Second)
+	assert.NoError(t, err)
+	orchestrator.cache = cache
+
+	return orchestrator
+}
+
+func newMockMapper(t *testing.T) mapper.Mapper {
+	bytes, err := ioutil.ReadFile("testdata/aggregation_rules.yaml") // key on request type
+	assert.NoError(t, err)
+
+	var config aggregationv1.KeyerConfiguration
+	err = yamlproto.FromYAMLToKeyerConfiguration(string(bytes), &config)
+	assert.NoError(t, err)
+
+	return mapper.NewMapper(&config)
+}
+
+func assertEqualResources(t *testing.T, got gcp.Response, expected v2.DiscoveryResponse, req gcp.Request) {
```

Comment:
In the future, you might want to share this and add it as a global testing utils function.

Reply:
Makes sense. We should refactor this out when/if there are multiple call sites.